src/typesystem.cpp

```diff
@@ -19,6 +19,16 @@ TypeSystemCore::TypeSystemCore() {
 		TypeFunctionData& b_data = m_type_functions[fb];
 
 		if (a_data.is_dummy) {
+
+			int new_argument_count = [&] {
+				if (a_data.argument_count == b_data.argument_count)
+					return a_data.argument_count;
+				else if (b_data.is_dummy || b_data.argument_count == -1)
+					return -1;
+				else
+					assert(0 && "Incompatible typefunc argument counts");
+			}();
```

Comment:
Wow this is unnecessary... but fancy. Perhaps even clearer. Hope it compiles to the same thing!